src/NuGet.Core/NuGet.Protocol/Utility/CachingUtility.cs

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace NuGet.Protocol
+{
+    public class CachingUtility
+    {
+        public static int BufferSize = 8192;
+
+        public static string ComputeHash(string value)
+        {
+            var trailing = value.Length > 32 ? value.Substring(value.Length - 32) : value;
+            byte[] hash;
+            using (var sha = SHA1.Create())
+            {
+                hash = sha.ComputeHash(Encoding.UTF8.GetBytes(value));
+            }
+
+            const string hex = "0123456789abcdef";
+            return hash.Aggregate("$" + trailing, (result, ch) => "" + hex[ch / 0x10] + hex[ch % 0x10] + result);
+        }
+
+        public static Stream TryReadCacheFile(TimeSpan maxAge, string cacheFile)
+        {
+            var fileInfo = new FileInfo(cacheFile);
+
+            if (fileInfo.Exists)
+            {
+                var age = DateTime.UtcNow.Subtract(fileInfo.LastWriteTimeUtc);
+                if (age < maxAge)
+                {
+                    var stream = new FileStream(
+                        cacheFile,
+                        FileMode.Open,
+                        FileAccess.Read,
+                        FileShare.Read | FileShare.Delete,
+                        BufferSize,
+                        useAsync: true);
+
+                    return stream;
+                }
+            }
+
+            return null;
+        }
+
+        public static bool IsFileAlreadyOpen(string filePath)
+        {
+            FileStream stream = null;
+
+            try
+            {
+                stream = new FileStream(filePath, FileMode.Open, FileAccess.ReadWrite, FileShare.None);
+            }
+            catch
+            {
+                return true;
+            }
+            finally
+            {
+                if (stream != null)
+                {
+                    stream.Dispose();
+                }
+            }
+
+            return false;
+        }
+
+        public static string RemoveInvalidFileNameChars(string value)
+        {
+            var invalid = Path.GetInvalidFileNameChars();
+            return new string(
```

Comment:
i am not sure what's going on here, explain please?

Reply:
It's a common utility used in both http caching and now plugin caching. 

Some characters such as >,|,< are not valid for a path, but can be used in urls. 

More info https://msdn.microsoft.com/en-us/library/system.io.path.getinvalidfilenamechars%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396.